src/storage/v2/replication/replication.cpp

```diff
@@ -0,0 +1,402 @@
+    case CONNECTION_FAILED:
+      return "CONNECTION_FAILED";
+    case COULD_NOT_BE_PERSISTED:
+      return "COULD_NOT_BE_PERSISTED";
+  }
+}
+}  // namespace
+
+storage::ReplicationState::ReplicationState(bool restore, std::filesystem::path durability_dir) {
+  if (restore) {
+    utils::EnsureDirOrDie(durability_dir / durability::kReplicationDirectory);
+    durability_ = std::make_unique<kvstore::KVStore>(durability_dir / durability::kReplicationDirectory);
+  }
+}
+
+void storage::ReplicationState::Reset() {
+  replication_server_.reset();
+  replication_clients_.WithLock([&](auto &clients) { clients.clear(); });
+}
+
+bool storage::ReplicationState::SetMainReplicationRole(storage::Storage *storage) {
+  // We don't want to generate new epoch_id and do the
+  // cleanup if we're already a MAIN
+  if (GetRole() == replication::ReplicationRole::MAIN) {
+    return false;
+  }
+
+  // Main instance does not need replication server
+  // This should be always called first so we finalize everything
+  replication_server_.reset(nullptr);
+
+  storage->EstablishNewEpoch();
+
+  if (ShouldStoreAndRestoreReplicationState()) {
+    // Only thing that matters here is the role saved as MAIN
+    auto data = replication::ReplicationStatusToJSON(
+        replication::ReplicationStatus{.name = replication::kReservedReplicationRoleName,
+                                       .ip_address = "",
+                                       .port = 0,
+                                       .sync_mode = replication::ReplicationMode::SYNC,
+                                       .replica_check_frequency = std::chrono::seconds(0),
+                                       .ssl = std::nullopt,
+                                       .role = replication::ReplicationRole::MAIN});
+
+    if (!durability_->Put(replication::kReservedReplicationRoleName, data.dump())) {
+      spdlog::error("Error when saving MAIN replication role in settings.");
+      return false;
+    }
+  }
+
+  SetRole(replication::ReplicationRole::MAIN);
+  return true;
+}
+
+bool storage::ReplicationState::AppendToWalDataDefinition(const uint64_t seq_num,
+                                                          durability::StorageGlobalOperation operation, LabelId label,
+                                                          const std::set<PropertyId> &properties,
+                                                          uint64_t final_commit_timestamp) {
+  bool finalized_on_all_replicas = true;
+  // TODO Should we return true if not MAIN?
+  if (GetRole() == replication::ReplicationRole::MAIN) {
+    replication_clients_.WithLock([&](auto &clients) {
+      for (auto &client : clients) {
+        client->StartTransactionReplication(seq_num);
+        client->IfStreamingTransaction(
+            [&](auto &stream) { stream.AppendOperation(operation, label, properties, final_commit_timestamp); });
+
+        const auto finalized = client->FinalizeTransactionReplication();
```

Comment:
Shouldn't we append TransactionEnd to stream before calling FinalizeTransactionReplication?

Reply:
Unknown, this refactor didn't change what was there before.
We will double check/validate this replication behavior later, but for now it is consistent with what was before + passes jepsen.